src/Microsoft.IdentityModel.Protocols.SignedHttpRequest/SignedHttpRequestHandler.cs

```diff
@@ -542,33 +542,55 @@ protected virtual async Task<SecurityToken> ValidateSignedHttpRequestPayloadAsyn
                 ValidateUClaim(jwtSignedHttpRequest, signedHttpRequestValidationContext);
 
-            if (signedHttpRequestValidationContext.SignedHttpRequestValidationParameters.ValidateP)
+            if (ShouldValidate(jwtSignedHttpRequest, validationParameters.ValidateP, ValidateIfPresent(validationParameters, SignedHttpRequestClaimTypes.P), SignedHttpRequestClaimTypes.P))
                 ValidatePClaim(jwtSignedHttpRequest, signedHttpRequestValidationContext);
 
-            if (signedHttpRequestValidationContext.SignedHttpRequestValidationParameters.ValidateQ)
+            if (ShouldValidate(jwtSignedHttpRequest, validationParameters.ValidateQ, ValidateIfPresent(validationParameters, SignedHttpRequestClaimTypes.Q), SignedHttpRequestClaimTypes.Q))
                 ValidateQClaim(jwtSignedHttpRequest, signedHttpRequestValidationContext);
 
-            if (signedHttpRequestValidationContext.SignedHttpRequestValidationParameters.ValidateH)
+            if (ShouldValidate(jwtSignedHttpRequest, validationParameters.ValidateH, ValidateIfPresent(validationParameters, SignedHttpRequestClaimTypes.H), SignedHttpRequestClaimTypes.H))
                 ValidateHClaim(jwtSignedHttpRequest, signedHttpRequestValidationContext);
 
-            if (signedHttpRequestValidationContext.SignedHttpRequestValidationParameters.ValidateB)
+            if (ShouldValidate(jwtSignedHttpRequest, validationParameters.ValidateB, ValidateIfPresent(validationParameters, SignedHttpRequestClaimTypes.B), SignedHttpRequestClaimTypes.B))
                 ValidateBClaim(jwtSignedHttpRequest, signedHttpRequestValidationContext);
 
             return jwtSignedHttpRequest;
         }
 
+        /// <summary>
+        /// Determine if validation of a claim should happen.
+        /// </summary>
+        /// <param name="jwtSignedHttpRequest">The request being considered to validate the claim on.</param>
+        /// <param name="validateClaim">Force validation to always occur.</param>
+        /// <param name="validateIfPresent">Validate if the claim is present.</param>
+        /// <param name="claimName">The name of the claim to validate.</param>
+        /// <returns>Whether the given claim should be validated.</returns>
+        internal virtual bool ShouldValidate(JsonWebToken jwtSignedHttpRequest, bool validateClaim, bool validateIfPresent, string claimName)
+        {
+            return validateClaim || (validateIfPresent && jwtSignedHttpRequest.TryGetClaim(claimName, out var claimValue) && claimValue != null);
+        }
+
+        private bool ValidateIfPresent(SignedHttpRequestValidationParameters validationParameters, string claim)
+        {
+            return validationParameters.ValidatePresentClaims &&
+                validationParameters.ClaimsValidatedWhenPresent != null &&
+                validationParameters.ClaimsValidatedWhenPresent.Contains(claim) == true;
```

Comment:
is `==true `needed ? Contains methods return true/false.

Reply:
No, it is not needed here.

Originally had this as `validationParameters.ClaimsValidatedWhenPresent?.Contains(claim) == true` to coerce the nullable bool to a bool. I ended up not doing that as I thought this was a little more clear

Reply:
Got it, Thanks.